extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,43 +15,118 @@
+export class AmpShareTracking extends AMP.BaseElement {
+  /**
+    * @return {boolean}
+    * @private
+    */
+  isExperimentOn_() {
+    return isExperimentOn(this.win, TAG);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
 
   /** @override */
   buildCallback() {
-    this.isExperimentOn_ = isExperimentOn(this.win, TAG);
-    if (!this.isExperimentOn_) {
+    if (!this.isExperimentOn_()) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    /** @private {string} */
+    this.vendorHref_ = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref_: ', this.vendorHref_);
+
+    /** @private {!Promise<!Object>} */
+    this.shareTrackingFragments_ = Promise.all([
+      this.getIncomingFragment_(),
+      this.getOutgoingFragment_(this.vendorHref_)]).then(results => {
+        return {
+          incomingFragment: results[0],
+          outgoingFragment: results[1],
+        }
+      });
+
   }
-}
 
-/**
- * ShareTrackingService processes the incoming and outcoming url fragment
- */
-export class ShareTrackingService {
-  constructor(window) {
-    this.win = window;
+  /**
+   * Get the incoming share-tracking fragment from the viewer
+   * @return {!Promise<string>}
+   * @private
+   */
+  getIncomingFragment_() {
+    return viewerFor(this.win).getFragment().then(fragment => {
+      if (!fragment || fragment.indexOf('.') != 0) {
+        return Promise.resolve();
+      }
+      const endIndex = fragment.indexOf('&');
+      if (endIndex != -1) {
+        return fragment.substr(1, (endIndex - 1));
+      } else {
+        return fragment.substr(1);
+      }
+    });
   }
-}
 
-/**
- * @param {!Window} window
- * @return {!ShareTrackingService}
- */
-export function installShareTrackingService(window) {
-  return getService(window, 'shareTrackingService', () => {
-    return new ShareTrackingService(window);
-  });
-}
+  /**
+   * Get an outgoing share-tracking fragment
+   * @param {string=} opt_vendorUrl
+   * @return {!Promise<string>}
+   * @private
+   */
+  getOutgoingFragment_(opt_vendorUrl) {
+    return this.getOutgoingFragmentFromVendor_(opt_vendorUrl).then(
+      fragment => (fragment || this.getOutgoingRandomFragment_())
+    );
+  }
 
-installShareTrackingService(AMP.win);
+  /**
+   * Get an outgoing share-tracking fragment from vendor
+   * by issueing a post request to the url the vendor provided
+   * @param {string=} opt_vendorUrl
+   * @return {!Promise<string>}
+   * @private
+   */
+  getOutgoingFragmentFromVendor_(opt_vendorUrl) {
+    if (!opt_vendorUrl) {
+      return Promise.resolve();
+    }
+    const postReq = {
+      method: 'POST',
+      credentials: 'include',
+      body: {},
+    };
+    return xhrFor(this.win).fetchJson(opt_vendorUrl, postReq).then(response => {
+      if (response.outgoingFragment) {
+        return response.outgoingFragment;
+      } else {
+        dev.warn(TAG, 'The response from [' + opt_vendorUrl +
```

Comment:
@rudygalfi in case of XHR response format error, or 404, do you think we should fallback to a random generator or just don't provide an outgoing fragment?

If we fallback, then in the analytics platform the publisher will see a mixture of his/her own fragment and random base64 fragment.